src/platform/datapath_raw.c

```diff
@@ -589,32 +589,63 @@ CXPLAT_THREAD_CALLBACK(CxPlatRouteResolutionWorkerThread, Context)
         CxPlatDispatchLockRelease(&Worker->Lock);
 
         while (!CxPlatListIsEmpty(&Operations)) {
+            MIB_IPNET_ROW2 IpnetRow = {0};
             CXPLAT_ROUTE_RESOLUTION_OPERATION* Operation =
                 CXPLAT_CONTAINING_RECORD(
                     CxPlatListRemoveHead(&Operations), CXPLAT_ROUTE_RESOLUTION_OPERATION, WorkerLink);
-            NETIO_STATUS Status =
-            Status = GetIpNetEntry2(&Operation->IpnetRow);
-            if (Status != ERROR_SUCCESS || Operation->IpnetRow.State <= NlnsIncomplete) {
-                Status =
-                    ResolveIpNetEntry2(&Operation->IpnetRow, NULL);
-                if (Status != 0) {
+            CXPLAT_DBG_ASSERT(
+                Operation->Route.State == RouteRefreshing || Operation->Route.State == RouteResolving);
+            CXPLAT_ROUTE NewRoute = Operation->Route;
+            QUIC_STATUS Status = CxPlatQueryRoute(Operation->Socket, &NewRoute, &IpnetRow);
+            if (Status == QUIC_STATUS_SUCCESS) {
+                if (!QuicAddrCompare(&NewRoute.LocalAddress, &Operation->Route.LocalAddress)) {
+                    //
+                    // We can't handle local address change here easily due to lack of full migration support.
+                    //
+                    Status = QUIC_STATUS_INVALID_STATE;
+                } else if (Operation->Route.State == RouteRefreshing &&
+                           memcmp(NewRoute.NextHopLinkLayerAddress,
+                                  Operation->Route.NextHopLinkLayerAddress,
+                                  sizeof(NewRoute.NextHopLinkLayerAddress)) == 0) {
+                    //
+                    // We are handling route refresh here. We will force neighbor discovery because next hop address
+                    // has not changed which implies this is the first time we are refreshing it. 
+                    //
+                    Status = QUIC_STATUS_PENDING;
+                }
+            }
+
+            if (Status == QUIC_STATUS_PENDING) {
```

Comment:
I was very confused about this at first because it looked like only the RouteRefreshing case would ever get into here. Now I see that you are returning PENDING from CxPlatQueryRoute when GetIpNetEntry2 doesn't return a result. Please don't use PENDING that way- PENDING is almost always used when a function kicks off some asynchronous work that you need to wait for completion of, which isn't the case here afaict.

Reply:
makes sense. let me see if we can find a better status code.